src/mprispropertiesadaptor.cpp

```diff
@@ -0,0 +1,239 @@
+    {"CanRaise", "canRaise()"},
+    {"HasTrackList", "hasTrackList()"},
+    {"Identity", "identity()"},
+    {"DesktopEntry", "desktopEntry()"},
+    {"SupportedUriSchemes", "supportedMimeTypes()"},
+    {"SupportedMimeTypes", "supportedUriSchemes()"},
+};
+
+QDBusVariant MprisPropertiesAdaptor::Get(const QString &interface_name, const QString &property_name)
+{
+    QMap<QString, QString> getMap;
+    QVariant result;
+
+    if (interface_name == QStringLiteral("org.mpris.MediaPlayer2.Player")) {
+        getMap = playerGetMap;
+    } else if (interface_name == QStringLiteral("org.mpris.MediaPlayer2")) {
+        getMap = serviceGetMap;
+    }
+
+    if (getMap.contains(property_name)) {
+        result = Get(getMap.value(property_name));
+    } else {
+        ReplyPropertyNotFoundError(interface_name, property_name);
+    }
+
+    return QDBusVariant(result);
+}
+
+QVariant MprisPropertiesAdaptor::Get(const QString &getter)
+{
+    QVariant result;
+    qDebug() << "MPRIS: Get called";
+
+    // Start by checking MprisPlayerPrivate methods
+    int pidx = m_playerPrivate->metaObject()->indexOfMethod(
+                getter.toLocal8Bit().data());
+
+    if (pidx != -1) {
+        const QMetaMethod method = m_playerPrivate->metaObject()->method(pidx);
+        Q_ASSERT(method.parameterCount() == 0);
+
+        const int returnType = method.returnType();
+        void * returnValue = QMetaType::create(returnType);
+        bool success = method.invoke(m_playerPrivate, Qt::DirectConnection,
+                                     QGenericReturnArgument(method.typeName(),
+                                                            returnValue));
+        if (success) {
+            // Conversion to QVariant will take a copy
+            result = QVariant(returnType, returnValue);
+        } else {
+            ReplyInternalError();
+        }
+        QMetaType::destroy(returnType, returnValue);
+    } else {
+        // Nothing in MprisPlayerPrivate so check MprisPlayer properties instead
+        pidx = m_playerPrivate->q_ptr->metaObject()->indexOfProperty(
+                    getter.toLocal8Bit().data());
+
+        if (pidx != -1) {
+            QMetaProperty mp = m_playerPrivate->q_ptr->metaObject()->property(pidx);
+            result = mp.read(m_playerPrivate->q_ptr);
+        } else {
+            ReplyInternalError();
+        }
+    }
+
+    return result;
+}
+
+QVariantMap MprisPropertiesAdaptor::GetAll(const QString &interface_name)
+{
+    qDebug() << "MPRIS: GetAll called: " << interface_name;
+    QVariantMap result;
+    QMap<QString, QString> getMap;
+
+    if (interface_name == QStringLiteral("org.mpris.MediaPlayer2.Player")) {
+        getMap = playerGetMap;
+    } else if (interface_name == QStringLiteral("org.mpris.MediaPlayer2")) {
+        getMap = serviceGetMap;
+    } else {
+        ReplyPropertyNotFoundError(interface_name, "");
+    }
+
+    for (const QString &property_name : getMap.keys()) {
```

Comment:
Using iterator would avoid the construction of the QList and value lookups.

Reply:
Yes, good idea. I've changed this too.